### PR TITLE
Refactor tensor_new.cpp to use TensorOptions instead of DispatchKey

### LINKED ISSUE #53544: Compute typeIdWithDefault and denseTypeIdWithDefault without reference to Backend
Currently they are implemented as

```
// NB: device_idx here is NOT a DeviceIndex, but index into PythonArgs
c10::DispatchKey typeIdWithDefault(PythonArgs& r, int64_t device_idx, c10::DispatchKey dispatch_key) {
  auto device_type = r.isNone(device_idx) ? computeDeviceType(dispatch_key) : r.device(device_idx).type();
  return backendToDispatchKey(backendToBackendOfDeviceType(dispatchKeyToBackend(dispatch_key), device_type));
}

// NB: device_idx here is NOT a DeviceIndex, but index into PythonArgs
c10::DispatchKey denseTypeIdWithDefault(PythonArgs& r, int64_t device_idx, c10::DispatchKey dispatch_key) {
  auto device_type = r.isNone(device_idx) ? computeDeviceType(dispatch_key) : r.device(device_idx).type();
  return backendToDispatchKey(toDense(backendToBackendOfDeviceType(dispatchKeyToBackend(dispatch_key), device_type)));
}     
```

This forces people who add new device types to also add a corresponding Backend for them. But the backend is just an incidental part of the computation, and we eventually turn it back to dispatch key. Do the calculation to dispatch key directly!

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Add some trivial meta functions for commonly used inplace operations
* #54127 Fix inaccurate dispatch tables
* #54079 Add Tensor::is_cpu, genericize TensorIterator
* **#54034 Refactor tensor_new.cpp to use TensorOptions instead of DispatchKey**

Fixes #53544

I had to touch a bunch of lines but the refactoring was fairly
mechanical.  Here's how it works.

The basic concept behind this PR is that tensor_new.cpp was previously
abusing DispatchKey when it actually meant TensorOptions.  The provided
DispatchKey argument to most of the constructor functions typically
comes from torch::tensors::get_default_dispatch_key();  it doesn't
really make sense for people to set the default dispatch key, but
this got grandfathered in due to the old API set_default_tensor_type
(where the "Type" concept got refactored into "DispatchKey" concept
over time).  See also #53124.  But the upshot is that, semantically,
what we refer to as the default dispatch key really is more like
torch.set_default_tensor_type(torch.Tensor) versus
torch.set_default_tensor_type(torch.cuda.Tensor): clearly the user
wants to do something about *construction* of the tensor, and
TensorOptions captures that exactly.

So, how exactly to translate from one to the other?
- Sources (things that used to PRODUCE DispatchKey)
  - Most top level functions take a DispatchKey as their argument.  I
    use the new function dispatchKeyToTensorOptions to convert it into
    a TensorOptions
  - typeIdWithDefault now produces a TensorOptions (probably could do
    with a rename, though I didn't)
- Sinks (things that used to CONSUME DispatchKey)
  - Previously, the function options() was typically used to convert the
    DispatchKey into a TensorOptions.  Now its replacement build_options
    just takes a TensorOptions and sets some extra fields on it.
    Irritatingly, I can't just replace
    `build_options(options, scalar_type, device)` with
    `options.dtype(scalar_type).device(device)` because the semantics
    are slightly different: if device is nullopt, we should preserve
    the usage of the device specified in options (what options.device()
    does is overwrite the device unconditionally; e.g., if device is
    nullopt, unset device from options)
  - The other major sink for DispatchKey was `internal_new_from_data`,
    but it turns out it only really extracts the device type from
    the dispatch key.  Now it just pulls out the device from
    TensorOptions.
- To actually do the translation of DispatchKey to TensorOptions, I
  introduce new functions dispatchKeyToLayout (replicating
  layout_from_backend--there are still a few uses of this function
  so I couldn't delete it) and dispatchKeyToDeviceType (replacing
  computeDeviceType)
- In all internal functions, whenever DispatchKey is taken as an argument,
  I instead take TensorOptions as an argument, and pass it along.
- Anywhere `legacyExtractDispatchKey(other.key_set())` equality was
  previously used, I now do `other.options().type_equal()`, which
  is the intended BC for doing "backend to backend" comparisons
- There are a few places in the sparse constructors where we allocated
  a tensor for values, and then read out the dispatch key from the
  result to allocate the keys.  As best as I can tell, this is totally
  equivalent to just passing in the options to both values and indices
  (the only difference is dtype, which is captured via a separate
  argument)

This refactor doesn't really go far enough: for example, there are now
functions that take both TensorOptions and ScalarType, when really
the TensorOptions can capture this all.  I kept it solely just
s/DispatchKey/TensorOptions/ to reduce the number of possible bugs;
also, a lot of this will be mooted by a proper fix to #53124.

Even with this limited refactor, the payoff is sweet.  I can delete:

- backendToCPU
- backendToXPU
- backendToCUDA
- backendToHIP
- backendToBackendOfDeviceType

The reason I can do this is because I can simply overwrite layout in TensorOptions
to do the conversion, rather than having to type out each backend case
explicitly.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27109509](https://our.internmc.facebook.com/intern/diff/D27109509)

### CODE REVIEW

**Comment by ezyang** [comment]
Apparently I broke `torch.DoubleTensor(torch.DoubleTensor([0.0]))`. Whacky!

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -45,87 +45,58 @@ namespace torch { namespace utils {
...
   tensor.set_(std::move(storage));
   return tensor;
 }
 
-Tensor new_with_tensor(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, const Tensor& other) {
-  TORCH_CHECK_TYPE(legacyExtractDispatchKey(other.key_set()) == dispatch_key, "expected ",
-                   toString(dispatch_key), " (got ", toString(legacyExtractDispatchKey(other.key_set())), ")");
+Tensor new_with_tensor(c10::TensorOptions options, at::ScalarType scalar_type, const Tensor& other) {
+  TORCH_CHECK_TYPE(other.options().type_equal(options), "expected ",
+                   options, " (got ", other.options(), ")");
```
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:99]: This isn't right, because `other.options()` also includes dtype

**Comment by ezyang** [comment]
> There are a few places in the sparse constructors where we allocated
a tensor for values, and then read out the dispatch key from the
result to allocate the keys. As best as I can tell, this is totally
equivalent to just passing in the options to both values and indices
(the only difference is dtype, which is captured via a separate
argument)

The CUDA failures suggest I was wrong here.

**Comment by ezyang** [comment]
This needs reviews!

File: torch/csrc/utils/tensor_new.cpp [hunk 2]
```diff
@@ -476,20 +438,25 @@ Tensor legacy_sparse_tensor_new(c10::DispatchKey dispatch_key, at::ScalarType sc
...
+    return new_with_sizes(options, scalar_type, r.deviceOptional(1), r.intlist(0));
   }
   throw std::runtime_error("new(): invalid arguments");
 }
 
 // NB: device_idx here is NOT a DeviceIndex, but index into PythonArgs
-c10::DispatchKey typeIdWithDefault(PythonArgs& r, int64_t device_idx, c10::DispatchKey dispatch_key) {
-  auto device_type = r.isNone(device_idx) ? computeDeviceType(dispatch_key) : r.device(device_idx).type();
-  return backendToDispatchKey(backendToBackendOfDeviceType(dispatchKeyToBackend(dispatch_key), device_type));
+c10::TensorOptions typeIdWithDefault(PythonArgs& r, int64_t device_idx, c10::DispatchKey dispatch_key) {
```
Comment (bhosmer) [inline torch/csrc/utils/tensor_new.cpp:447]: This really is overdue for a rename but yeah there's no obvious upgrade. What about just optionsWithDefaultDevice or something?

File: torch/csrc/utils/tensor_new.cpp [hunk 3]
```diff
@@ -476,20 +438,25 @@ Tensor legacy_sparse_tensor_new(c10::DispatchKey dispatch_key, at::ScalarType sc
...
+  auto options = dispatchKeyToTensorOptions(dispatch_key);
+  if (!r.isNone(device_idx)) {
+    // TODO: This line doesn't seem to be exercised at all in tests
+    options = options.device(r.device(device_idx).type());
```
Comment (bhosmer) [inline torch/csrc/utils/tensor_new.cpp:451]: Doesn't this still have the problem you described in the "janky" note? Index gets scraped off in the same way as before I think.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (bhosmer) [reply torch/csrc/utils/tensor_new.cpp:88]: a) I'm not sure which thing isn't right here b) regardless, is it worth putting in a comment?

File: torch/csrc/utils/tensor_new.cpp [hunk 4]
```diff
@@ -580,49 +548,70 @@ Tensor legacy_tensor_new(c10::DispatchKey dispatch_key, at::ScalarType scalar_ty
...
+//  torch.sparse_coo_tensor(([0, 1],), self.empty(2, 0).cuda(), (4, 0))
+//
+// Sparse tensors require both indices and values to live on the same device.
+// If values lives on CUDA, we can infer where the indices should live, and
+// should accept even ordinary index sequences (and just make sure we write them
+// into the correct device).  values is the ONLY way we know that the index
+// tensor should go to CUDA, so we have to get the information in somehow.
+//
+// This code is kind of jank.  For one, the dtype in options is silently ignored
+// by internal_new_from_data.  Also, in classic janky code style, it used to
```
Comment (bhosmer) [inline torch/csrc/utils/tensor_new.cpp:610]: This is a choice we're making in adopting the new scheme though, right? The relevant subset of stuff carried by an options argument will be context-dependent.

**Review by bhosmer — APPROVED** [review]
Nice cleanup - one nontrivial thing inline (if I'm reading the code right)

File: torch/csrc/utils/tensor_new.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/utils/tensor_new.cpp:451]: Yes, I think you're right. It's too bad I can't tell if it matters or not cuz it's not exercised by tests LOL

File: torch/csrc/utils/tensor_new.cpp (hunk 4 above)
Comment (ezyang) [reply torch/csrc/utils/tensor_new.cpp:610]: Not... really? Previously the code accepted DispatchKey, but ignored everything besides the Device associated with that DispatchKey. Options is a bigger type, but the mismatch in size was always there in the code.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/utils/tensor_new.cpp:88]: This comment is out of date; the code is updated!

**Comment by ezyang** [comment]
Gonna see if I can land this stack ready to go (if it lands, I'll fix these in follow up), otherwise I'll address these comments

**Pull request closed without merging**